### ISSUE TITLE
Test modules and examples with Terraform 0.12.x

### ISSUE THREAD

**Issue by ketzacoatl** [fpco/terraform-aws-foundation#199]
We haven't yet used Terraform 0.12.x on the repository, and there are likely to be 0.12.x-specific updates we need to review and make. We do not need to maintain compatibility with 0.11.x, we can say version X of the modules supports 0.11.x while the new releases do not.

## Requirements

* Use the latest stable release of Terraform 0.12.x
* This could take for ever if we wanted to review everything, so instead, pick high-traffic examples and modules and go for breadth of review more than completeness.
* Running Terraform init/plan should be enough to exercise the details and confirm if there are issues to update, but we may need to do a full `apply` on the example to do that testing, that's ok.
* When we find a particular type issue, review the rest of the repo for that same type of issue and update.

## Examples Env to Test

* [X] `asg-lifecycle-hooks`
* [ ] `cloud-dev-workspace`
* [X] `gitlab-ha`
* [ ] `gitlab-simple-ha`
* [ ] `kops-with-vpc`
* [ ] `kube-stack-private`
* [ ] `kube-stack-public`
* [ ] `load-asg`
* [ ] `nexus-asg`
* [ ] `s3-full-access-policy`
* [X] `s3-remote-state`
* [ ] `terraform-vpc`
* [ ] `vpc-scenario-1`
* [ ] `vpc-scenario-2-nat-instance`
* [ ] `vpc-scenario-2-nat-instances-per-az`
* [X] `vpc-scenario-2`
* [ ] `vpc-scenario-3`
* [ ] `vpc-scenario-peering`
* [ ] `vpn-gateway`

**Comment by Magicloud** [comment]
- bind-server/
```
Error: Unsupported block type

  on main.tf line 67, in resource "null_resource" "bind":
  67:   triggers {

Blocks of type "triggers" are not expected here. Did you mean to define
argument "triggers"? If so, use the equals sign to assign it a value.
```
- consul-agent-generic-init/
```
Error: Unsupported block type

  on main.tf line 47, in resource "template_file" "generic_init":
  47:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- consul-demo-server/
```
Error: Unsupported block type

  on ../init-snippet-attach-ebs-volume/main.tf line 55, in data "template_file" "init_snippet":
  55:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- consul-leaders-generic-init/
```
Error: Unsupported block type

  on main.tf line 17, in resource "template_file" "generic_init":
  17:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- cross-account-role/
> to confirm
- dnsmasq-server/
```
Error: Unsupported block type

  on main.tf line 45, in resource "null_resource" "dnsmasq":
  45:   triggers {

Blocks of type "triggers" are not expected here. Did you mean to define
argument "triggers"? If so, use the equals sign to assign it a value.
```
- ec2-auto-recover-instances/
```
Error: Unsupported block type

  on main.tf line 7, in data "template_file" "cnt":
   7:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- ec2-nat-instances/
```
Error: Unsupported block type

  on ../ec2-auto-recover-instances/main.tf line 7, in data "template_file" "cnt":
   7:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.


Error: Unsupported block type

  on ../init-snippet-exec/main.tf line 17, in data "template_file" "init_snippet":
  17:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.


Error: Unsupported block type

  on ../init-snippet-hostname-simple/main.tf line 35, in data "template_file" "init_snippet":
  35:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- elasticsearch/
```
Error: Attribute redefined

  on master-nodes.tf line 43, in resource "aws_launch_configuration" "master-node-lc":
  43:   security_groups      = ["${aws_security_group.transport-sg.id}"]

The argument "security_groups" was already set at master-nodes.tf:42,3-18.
Each argument may be set only once.
```
- elk-stack/
```
Error: Attribute redefined

  on ../elasticsearch/master-nodes.tf line 43, in resource "aws_launch_configuration" "master-node-lc":
  43:   security_groups      = ["${aws_security_group.transport-sg.id}"]

The argument "security_groups" was already set at
../elasticsearch/master-nodes.tf:42,3-18. Each argument may be set only once.


Error: Missing key/value separator

  on ../logstash/iam.tf line 51, in data "aws_iam_policy_document" "logstash-role":
  47: 
  48: 
  49: 
  51:     principal {

Expected an equals sign ("=") to mark the beginning of the attribute value.


Error: Attribute redefined

  on ../elasticsearch/master-nodes.tf line 43, in resource "aws_launch_configuration" "master-node-lc":
  43:   security_groups      = ["${aws_security_group.transport-sg.id}"]

The argument "security_groups" was already set at
../elasticsearch/master-nodes.tf:42,3-18. Each argument may be set only once.


Error: Missing key/value separator

  on ../logstash/iam.tf line 51, in data "aws_iam_policy_document" "logstash-role":
  47: 
  48: 
  49: 
  51:     principal {

Expected an equals sign ("=") to mark the beginning of the attribute value.
```
- ha-management-cluster/
```
Error: Unsupported block type

  on ../consul-agent-generic-init/main.tf line 47, in resource "template_file" "generic_init":
  47:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-attach-ebs-volume/
```
Error: Unsupported block type

  on main.tf line 55, in data "template_file" "init_snippet":
  55:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-config-consul-leader/
```
Error: Unsupported block type

  on main.tf line 47, in data "template_file" "init_snippet":
  47:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-config-upstart-consul/
```
Error: Unsupported block type

  on main.tf line 49, in data "template_file" "init_snippet":
  49:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-consul-agent/
```
Error: Unsupported block type

  on main.tf line 12, in data "template_file" "init_snippet":
  12:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-consul-leader/
```
Error: Unsupported block type

  on main.tf line 16, in data "template_file" "init_snippet":
  16:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-consul-template/
```
Error: Unsupported block type

  on main.tf line 47, in data "template_file" "init_snippet":
  47:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-curator/
```
Error: Unsupported block type

  on main.tf line 20, in data "template_file" "curator-setup":
  20:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-exec/
```
Error: Unsupported block type

  on main.tf line 17, in data "template_file" "init_snippet":
  17:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-hostname/
```
Error: Unsupported block type

  on main.tf line 52, in data "template_file" "init_snippet":
  52:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-hostname-simple/
```
Error: Unsupported block type

  on main.tf line 35, in data "template_file" "init_snippet":
  35:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-install-awscli/
> to confirm
- init-snippet-install-consul/
```
Error: Invalid variable name

  on main.tf line 17, in variable "version":
  17: variable "version" {

The variable name "version" is reserved due to its special meaning inside
module blocks.
```
- init-snippet-install-ops/
```
Error: Invalid variable name

  on main.tf line 9, in variable "version":
   9: variable "version" {

The variable name "version" is reserved due to its special meaning inside
module blocks.
```
- init-snippet-nexus/
```
Error: Unsupported block type

  on main.tf line 42, in data "template_file" "init_snippet":
  42:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-nomad-agent/
```
Error: Unsupported block type

  on main.tf line 16, in data "template_file" "init_snippet":
  16:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-prometheus/
```
Error: Unsupported block type

  on main.tf line 45, in data "template_file" "init_snippet":
  45:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- init-snippet-write-bootstrap-pillar/
```
Error: Unsupported block type

  on main.tf line 32, in data "template_file" "init_snippet":
  32:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- kibana/
```
Error: Unsupported block type

  on main.tf line 55, in data "template_file" "kibana-setup":
  55:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- logstash/
```
Error: Missing key/value separator

  on iam.tf line 51, in data "aws_iam_policy_document" "logstash-role":
  47: 
  48: 
  49: 
  51:     principal {

Expected an equals sign ("=") to mark the beginning of the attribute value.
```
- persistent-ebs-volumes/
```
Error: Unsupported block type

  on main.tf line 31, in data "template_file" "volume_mount_snippets":
  31:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- prometheus-server/
```
Error: Unsupported block type

  on ../init-snippet-attach-ebs-volume/main.tf line 55, in data "template_file" "init_snippet":
  55:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.


Error: Unsupported block type

  on ../init-snippet-hostname/main.tf line 52, in data "template_file" "init_snippet":
  52:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.


Error: Unsupported block type

  on ../init-snippet-prometheus/main.tf line 45, in data "template_file" "init_snippet":
  45:   vars {

Blocks of type "vars" are not expected here. Did you mean to define argument
"vars"? If so, use the equals sign to assign it a value.
```
- single-node-asg/
> to confirm
- ssh-sg/
> to confirm
- vpc-legacy/
```
Error: Unsupported argument

  on main.tf line 37, in module "subnets":
  37:   public_subnet_cidrs  = "${var.public_subnet_cidrs}"

An argument named "public_subnet_cidrs" is not expected here.


Error: Unsupported argument

  on main.tf line 38, in module "subnets":
  38:   private_subnet_cidrs = "${var.private_subnet_cidrs}"

An argument named "private_subnet_cidrs" is not expected here.
```

**Comment by Magicloud** [comment]
terraform 0.11.14 provides 0.12checklist command to verify the configurations. But from my understanding of https://www.terraform.io/upgrade-guides/0-12.html, there are also behavior changes besides syntax changes.

I am working on making modules pass checklist first.

**Comment by Magicloud** [comment]
ebs-snapshot-iam/
After analyzing this configuration and working directory, we have identified some necessary steps that we recommend you take before upgrading to Terraform v0.12:

- [ ] Terraform couldn't reach the Terraform Registry (at `registry.terraform.io`) to determine whether current provider plugins are v0.12-compatible.
  
  In general, we recommend upgrading to the latest version of each provider before upgrading to Terraform v0.12.

Taking these steps before upgrading to Terraform v0.12 will simplify the upgrade process by avoiding syntax errors and other compatibility problems.

init-snippet-install-consul/
After analyzing this configuration and working directory, we have identified some necessary steps that we recommend you take before upgrading to Terraform v0.12:

- [ ] The variable name "version" is now reserved.
  
  Terraform 0.12 reserves certain variable names that have (or will have in future) a special meaning when used in a "module" block. The name of this variable must be changed before upgrading to v0.12. This will unfortunately be a breaking change for any user of this module, requiring a major release to indicate that.
  For more information on the reserved variable names, see the documentation at https://www.terraform.io/docs/configuration/variables.html .

Taking these steps before upgrading to Terraform v0.12 will simplify the upgrade process by avoiding syntax errors and other compatibility problems.

init-snippet-install-ops/
After analyzing this configuration and working directory, we have identified some necessary steps that we recommend you take before upgrading to Terraform v0.12:

- [ ] The variable name "version" is now reserved.
  
  Terraform 0.12 reserves certain variable names that have (or will have in future) a special meaning when used in a "module" block. The name of this variable must be changed before upgrading to v0.12. This will unfortunately be a breaking change for any user of this module, requiring a major release to indicate that.
  For more information on the reserved variable names, see the documentation at https://www.terraform.io/docs/configuration/variables.html .

Taking these steps before upgrading to Terraform v0.12 will simplify the upgrade process by avoiding syntax errors and other compatibility problems.

vpc-legacy/

Error: module "subnets": "public_subnet_cidrs" is not a valid argument



Error: module "subnets": "private_subnet_cidrs" is not a valid argument



Error: module "subnets": missing required argument "cidr_blocks"



Error: resource 'aws_route_table_association.public-rta' config: "public_ids" is not a valid output for module "subnets"



Error: output 'private_subnet_ids': "private_ids" is not a valid output for module "subnets"



Error: module 'nat-gateways': "public_ids" is not a valid output for module "subnets"



Error: module 'nat-gateways': "private_ids" is not a valid output for module "subnets"



Error: output 'public_subnet_ids': "public_ids" is not a valid output for module "subnets"



vpc-scenario-1/
After analyzing this configuration and working directory, we have identified some necessary steps that we recommend you take before upgrading to Terraform v0.12:

- [ ] Provider "aws" may need to be upgraded to a newer version that supports Terraform 0.12. (Request for supported version information failed: Get https://registry.terraform.io/v1/providers/-/aws/versions: net/http: TLS handshake timeout.)

# Module `"public-gateway"`

# Module `"public-subnets"`

# Module `"vpc"`

Taking these steps before upgrading to Terraform v0.12 will simplify the upgrade process by avoiding syntax errors and other compatibility problems.

**Comment by Magicloud** [comment]
nexus-asg/

Error: module "snasg": "name" is not a valid argument



Error: module "snasg": "key_file" is not a valid argument



Error: module "snasg": "az" is not a valid argument



Error: module "snasg": missing required argument "name_prefix"

**Comment by Magicloud** [comment]
- elasticsearch/
```
Error: Unrecognized block type

  on data-nodes.tf line 278, in data "aws_iam_policy_document" "source_policy":
 278:     principal {

Blocks of type "principal" are not expected in
data.aws_iam_policy_document.source_policy.statement.


Error: Unrecognized attribute name

  on data-nodes.tf line 283, in data "aws_iam_policy_document" "source_policy":
 283:     action = ["sts:AssumeRole"]

No attribute named "action" is expected in
data.aws_iam_policy_document.source_policy.statement.


Error: Unrecognized attribute name

  on data-nodes.tf line 288, in data "aws_iam_policy_document" "master-node-role":
 288:   source_policy_json = "${data.aws_iam_policy_document.source_policy.json}"

No attribute named "source_policy_json" is expected in
data.aws_iam_policy_document.master-node-role.


Error: Unrecognized attribute name

  on data-nodes.tf line 292, in data "aws_iam_policy_document" "data-node-role":
 292:   source_policy_json = "${data.aws_iam_policy_document.source_policy.json}"

No attribute named "source_policy_json" is expected in
data.aws_iam_policy_document.data-node-role.


Error: Unrecognized attribute name

  on data-nodes.tf line 298, in data "aws_iam_policy_document" "ec2-discovery-policy":
 298:     resource = ["*"]

No attribute named "resource" is expected in
data.aws_iam_policy_document.ec2-discovery-policy.statement.


Error: Unrecognized attribute name

  on data-nodes.tf line 299, in data "aws_iam_policy_document" "ec2-discovery-policy":
 299:     action   = ["ec2:DescribeInstances"]

No attribute named "action" is expected in
data.aws_iam_policy_document.ec2-discovery-policy.statement.
```
- elk-stack/ (uses elasticsearch which is broken)
```
Error: error resolving providers:

- provider.aws: no suitable version installed
  version requirements: "(any version)"
  versions installed: none
```
- logstash/
```
Error: Unrecognized block type

  on iam.tf line 51, in data "aws_iam_policy_document" "logstash-role":
  51:     principal {

Blocks of type "principal" are not expected in
data.aws_iam_policy_document.logstash-role.statement.


Error: Unrecognized attribute name

  on iam.tf line 56, in data "aws_iam_policy_document" "logstash-role":
  56:     action = ["sts:AssumeRole"]

No attribute named "action" is expected in
data.aws_iam_policy_document.logstash-role.statement.
```

**Comment by ketzacoatl** [comment]
I'm not sure if you have a WIP local branch for this, but I am using `terraform 0.12upgrade` in a for loop with `terraform init` to run thru each module and example and run this upgrade. So far, this has worked well in practice, to get a bunch of the code auto-updated. I'll push a branch with Terraform's changes to the modules and examples, which we can then run more tests on to find the specific problems that remain.

**Comment by Magicloud** [comment]
Yes, I have. `0.12upgrade` could not solve all the issues. Many manual works have to be done. And I am on it.

**Comment by ketzacoatl** [comment]
Here is the loop I am running:

```
modules$ for m in `ls` ; do echo updating $m ; cd $m && cp -r ../.terraform ./ && cp ../providers.tf ./ && terraform-0.12.2 init && terraform-0.12.2 0.12upgrade -yes ; cd .. ; done
```
where `.terraform` has the `plugins` after `terraform init` for a module with `aws`, `template`, and `local` providers (so there is no need to download those), and `providers.tf` has an entry for the `aws` provider (to set the `region`).

**Comment by ketzacoatl** [comment]
Have you run the `0.12upgrade` in all or some modules? If you could push that branch, it would be helpful, as I could then use it for some of the R&D I'm doing in parallel (I mostly using VPC examples w/ ASG).

**Comment by Magicloud** [comment]
Sure. A minute.

**Comment by ketzacoatl** [comment]
Great, that's perfect, thanks!

**Comment by Magicloud** [comment]
What would the "count" be here?

modules/vpc-scenario-3/main.tf
`98:       "Name" = "${var.name_prefix}-private-vpn-${format("%02d", count.index)}"`
The "count" object can be used only in "resource" and "data" blocks, and only
when the "count" argument is set.

**Comment by Magicloud** [comment]
```
vpc-scenario-3/

Error: Reference to "count" in non-counted context

  on main.tf line 98, in resource "aws_route_table" "private-vpn":
  98:       "Name" = "${var.name_prefix}-private-vpn-${format("%02d", count.index)}"

The "count" object can be used only in "resource" and "data" blocks, and only
when the "count" argument is set.
```
```
setup-meta-infrastructure/

Error: Invalid value for module argument

  on policies.tf line 56, in module "assume-admin-role-policy":
  56:   account_ids = [
  57:     data.aws_caller_identity.current.account_id,
  58:     var.admin_control_account_ids,
  59:   ]

The given value is not suitable for child module variable "account_ids"
defined at ../cross-account-assume-role-policy/main.tf:18,1-23: element 1:
string required.


Error: Invalid value for module argument

  on policies.tf line 145, in module "assume-power-user-role-policy":
 145:   account_ids = [
 146:     data.aws_caller_identity.current.account_id,
 147:     var.power_user_control_account_ids,
 148:   ]

The given value is not suitable for child module variable "account_ids"
defined at ../cross-account-assume-role-policy/main.tf:18,1-23: element 1:
string required.


Error: Invalid value for module argument

  on roles.tf line 9, in module "admin-role":
   9:   trust_account_ids = [
  10:     data.aws_caller_identity.current.account_id,
  11:     var.trust_account_ids,
  12:   ]

The given value is not suitable for child module variable "trust_account_ids"
defined at ../cross-account-role/main.tf:9,1-29: element 1: string required.


Error: Invalid value for module argument

  on roles.tf line 19, in module "power-user-role":
  19:   trust_account_ids = [
  20:     data.aws_caller_identity.current.account_id,
  21:     var.trust_account_ids,
  22:   ]

The given value is not suitable for child module variable "trust_account_ids"
defined at ../cross-account-role/main.tf:9,1-29: element 1: string required.
```
```
persistent-ebs-volumes/

Error: Reference to "count" in non-counted context

  on iam.tf line 26, in data "aws_iam_policy_document" "ebs-volume-policy":
  26:       "arn:${data.aws_partition.current.partition}:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:volume/${element(aws_ebs_volume.volumes.*.id, count.index)}",

The "count" object can be used only in "resource" and "data" blocks, and only
when the "count" argument is set.
```
```
elasticsearch/

Error: Reference to undeclared resource

  on data-nodes.tf line 348, in data "aws_iam_policy_document" "master-node-role":
 348: source_json = data.aws_iam_policy_document.source.json

A data resource "aws_iam_policy_document" "source" has not been declared in
the root module.


Error: Reference to undeclared resource

  on data-nodes.tf line 352, in data "aws_iam_policy_document" "data-node-role":
 352: source_json = data.aws_iam_policy_document.source.json

A data resource "aws_iam_policy_document" "source" has not been declared in
the root module.


Error: Reference to "count" in non-counted context

  on ../persistent-ebs-volumes/iam.tf line 26, in data "aws_iam_policy_document" "ebs-volume-policy":
  26:       "arn:${data.aws_partition.current.partition}:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:volume/${element(aws_ebs_volume.volumes.*.id, count.index)}",

The "count" object can be used only in "resource" and "data" blocks, and only
when the "count" argument is set.


Error: Reference to "count" in non-counted context

  on ../persistent-ebs-volumes/iam.tf line 26, in data "aws_iam_policy_document" "ebs-volume-policy":
  26:       "arn:${data.aws_partition.current.partition}:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:volume/${element(aws_ebs_volume.volumes.*.id, count.index)}",

The "count" object can be used only in "resource" and "data" blocks, and only
when the "count" argument is set.
```
```
cross-account-assume-role-policy/

Error: Incorrect attribute value type

  on main.tf line 54, in data "aws_iam_policy_document" "assume-role":
  54:     resources = [

Inappropriate value for attribute "resources": element 0: string required.
```
```
bind-server/

Error: Incorrect attribute value type

  on main.tf line 112, in resource "null_resource" "bind":
 112:     inline = [

Inappropriate value for attribute "inline": element 8: string required.
```

**Comment by ketzacoatl** [comment]
Can you refer to the code in the repo by URL? I need to see the larger context.

**Comment by Magicloud** [comment]
You mean the error message? No worries. I am just making a log. Fixing on the way.

**Comment by ketzacoatl** [comment]
There was a separate comment I was referring to, sorry, I thought you were asking a question.

**Comment by Magicloud** [comment]
I see what you meant. The count usage. Also listed in https://github.com/fpco/terraform-aws-foundation/issues/199#issuecomment-504666750. I look through the code and fixed. But I do need a review on that. I am validating. Later will push the branch.

**Comment by Magicloud** [comment]
Pushed changes in modules.

**Comment by Magicloud** [comment]
**High priority reviews**
- [ ] vpc-scenario-2
- [ ] ami-ubuntu
- [ ] security-group-base
- [ ] single-port-sg
- [ ] open-egress-sg
- [ ] asg

**Comment by Magicloud** [comment]
Running `terraform plan` pops more errors for examples/.

**Comment by Magicloud** [comment]
Cannot `plan` _vpc-scenario-2_.
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.ubuntu-xenial-ami.data.aws_partition.current: Refreshing state...
data.aws_availability_zones.available: Refreshing state...
module.ubuntu-xenial-ami.data.aws_ami.ubuntu: Refreshing state...

Error: Error fetching Availability Zones: UnauthorizedOperation: You are not authorized to perform this operation.
	status code: 403, request id: dbb219f8-a2d4-4241-87a2-fb7be6ffcdf2

  on main.tf line 51, in data "aws_availability_zones" "available":
  51: data "aws_availability_zones" "available" {



Error: UnauthorizedOperation: You are not authorized to perform this operation.
	status code: 403, request id: 789a4a32-d5b9-4850-8927-f48c5c077951

  on ../../modules/ami-ubuntu/main.tf line 77, in data "aws_ami" "ubuntu":
  77: data "aws_ami" "ubuntu" {
```

**Comment by Magicloud** [comment]
A confusing part of **init-snippet-gitlab-docker**.
https://github.com/fpco/terraform-aws-foundation
```
variable "config_elb" { # line 82
  default = true # apparently this var is of type boolean.
}
data "template_file" "omnibus_config_elb" { # line 93
  count = var.config_elb # count expects number.
}
data "template_file" "omnibus_config_eip" { # line 112
 count = 1 - var.config_elb # count expects number.
}
```

**Comment by Magicloud** [comment]
@ketzacoatl Please help review and test the count change. It is a change I am not certian of.
https://github.com/fpco/terraform-aws-foundation/commit/83639dbbd0e950c6b51da5e808874c9dbd8b7ded

**Comment by ketzacoatl** [comment]
On the surface, yea, that looks right.

**Comment by Magicloud** [comment]
Since terraform may mistake an unclearified type. I am putting types to all variables. Following are the files I need to work.

./asg-lifecycle/variables.tf
./cluster-network/variables.tf
./consul-agent-generic-init/variables.tf
./consul-agent-sg/main.tf
./consul-cluster/variables.tf
./consul-demo-server/variables.tf
./consul-leader-sg/main.tf
./consul-leader-wan-sg/main.tf
./consul-leaders-generic-init/variables.tf
./consul-leaders/variables.tf
./ha-management-cluster/variables.tf
./init-snippet-attach-ebs-volume/main.tf
./init-snippet-config-consul-leader/main.tf
./init-snippet-config-upstart-consul/main.tf
./init-snippet-consul-agent/variables.tf
./init-snippet-consul-leader/variables.tf
./init-snippet-consul-template/main.tf
./init-snippet-curator/main.tf
./init-snippet-exec/main.tf
./init-snippet-gitlab-docker/main.tf
./init-snippet-hostname-simple/main.tf
./init-snippet-hostname/main.tf
./init-snippet-install-awscli/main.tf
./init-snippet-install-ops/main.tf
./init-snippet-nexus/main.tf
./init-snippet-nomad-agent/variables.tf
./init-snippet-prometheus/main.tf
./init-snippet-write-bootstrap-pillar/main.tf
./kubernetes-vpc/variables.tf
./multi-volumes/main.tf
./nomad-agent-sg/main.tf
./nomad-agent-worker-ports-sg/main.tf
./nomad-server-sg/main.tf
./persistent-ebs-volumes/variables.tf
./prometheus-server/tests/main.tf
./s3-full-access-policy/main.tf
./s3-remote-state/main.tf
./ssh-sg/main.tf

**Comment by ketzacoatl** [comment]
@Magicloud what do you recommend as our next steps? How much validation should we do before merging?

**Comment by Magicloud** [comment]
I won't recommending merging soon. But I am OK to let other people know this branch and give it a try.

On the other hand, I suggest to open another project as to add unit test of each module. At least get them pass `plan` with no human interfere. So we can use CI to verify changes in the future.

**Comment by ketzacoatl** [comment]
`vpc-scenario-2` deployed cleanly for me. I will test a few other env out as well.

**Comment by Magicloud** [comment]
Assign back to @ketzacoatl to allocate testing resource.

**Comment by ketzacoatl** [comment]
@AronT-TLV please be vocal in here when you choose an example, so other engineers running tests can choose other examples to test and distribute our work. I'll add a list to the description of the PR.

**Comment by ketzacoatl** [comment]
To the ticket description, I've added a checklist with the list of examples, and I've checked off the examples I tested last week.

**Comment by ketzacoatl** [comment]
The branch has been merged and `v0.9.0` has been released.
